File: runtime/schema/python/superschematic_schema_runtime/parse/defaults.py
"""Default-value parsing.

``FieldDef.default_value`` holds the schema-declared default as a string; this module
converts it into the target primitive when the field is absent during parse.
"""

from __future__ import annotations

import math
from typing import Any

from ..validation.types import FieldDef, ScalarDef


def primitive_for_builtin(name: str) -> str:
    if name in ("Int", "Float", "Boolean"):
        return name
    if name in ("String", "ID"):
        return "String"
    return ""


def apply_default(field: FieldDef, scalar: ScalarDef | None) -> tuple[Any, bool]:
    """Return ``(value, ok)`` for ``field``'s default.

    ``ok=False`` means the default string was not parseable for the declared
    type and the caller should emit a ``{"validator": "default"}`` error.
    """
    raw = field.default_value
    if raw is None:
        return None, False
    primitive = scalar.primitive if scalar is not None else primitive_for_builtin(
        field.type_ref.name
    )

    if primitive == "Int":
        try:
            parsed = float(raw)
        except ValueError:
            return None, False
        if not math.isfinite(parsed):
            return None, False
        return int(parsed), True
    if primitive == "Float":
        try:
            parsed = float(raw)
        except ValueError:
            return None, False
        if not math.isfinite(parsed):
            return None, False
        return parsed, True
    if primitive == "Boolean":
        if raw in ("true", "1"):
            return True, True
        if raw in ("false", "0"):
            return False, True
        return None, False
    return raw, True


__all__ = ["apply_default", "primitive_for_builtin"]

File: runtime/schema/typescript/src/runtime/parse/defaults.ts
/**
 * Default-value parsing. FieldDef.defaultValue holds the schema-declared default
 * as a string; this parses it into the appropriate primitive when the field is
 * absent during parse. Defaults on object-typed fields and arrays are NOT applied
 * here -- the walker skips applyDefault for those kinds.
 */

import type { FieldDef, ScalarDef } from '../validation/types';

export function applyDefault(
  field: FieldDef,
  scalar: ScalarDef | null
): [unknown, boolean] {
  if (field.defaultValue === null || field.defaultValue === undefined) {
    return [null, false];
  }
  const raw = field.defaultValue;
  const primitive = scalar
    ? scalar.primitive
    : primitiveForBuiltin(field.typeRef.name);

  switch (primitive) {
    case 'Int': {
      const parsed = Number(raw);
      if (!Number.isFinite(parsed)) {
        return [null, false];
      }
      return [Math.trunc(parsed), true];
    }
    case 'Float': {
      const parsed = Number(raw);
      if (!Number.isFinite(parsed)) {
        return [null, false];
      }
      return [parsed, true];
    }
    case 'Boolean': {
      if (raw === 'true' || raw === '1') return [true, true];
      if (raw === 'false' || raw === '0') return [false, true];
      return [null, false];
    }
    default:
      return [raw, true];
  }
}

export function primitiveForBuiltin(name: string): string {
  switch (name) {
    case 'Int':
      return 'Int';
    case 'Float':
      return 'Float';
    case 'Boolean':
      return 'Boolean';
    case 'String':
    case 'ID':
      return 'String';
    default:
      return '';
  }
}
